xdsl/utils/parse_pipeline.py

```diff
@@ -0,0 +1,227 @@
+
+
+_PassArgTypes = list[str | int | bool | float]
+
+
+def parse_pipeline(
+    pipeline_spec: str,
+) -> Iterator[tuple[str, dict[str, _PassArgTypes]]]:
+    """
+    This takes a pipeline string and gives a representation of
+    the specification.
+
+    Each pass is represented by a tuple of:
+     - name: the name of the pass as string
+     - args: a dictionary, where each value is zero or more
+            of (str | bool | float | int)
+    """
+    lexer = PipelineLexer(pipeline_spec)
+
+    while True:
+        name = lexer.lex()
+        if name.kind is Kind.EOF:
+            return
+        if name.kind is not Kind.IDENT:
+            raise PassPipelineParseError(name, "Expected pass name here")
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            yield name.span.text, dict()
+            return
+
+        if delim.kind is Kind.COMMA:
+            yield name.span.text, dict()
+            continue
+
+        if delim.kind is not Kind.L_BRACE:
+            raise PassPipelineParseError(
+                delim, "Expected a comma or pass arguments here"
+            )
+
+        yield name.span.text, _parse_pass_args(lexer)
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            return
+        if delim.kind is not Kind.COMMA:
+            raise PassPipelineParseError(
+                delim, "Expected a comma after pass argument dict here"
+            )
+
+
+def _parse_pass_args(lexer: PipelineLexer):
+    args: dict[str, _PassArgTypes] = dict()
+
+    while True:
+        name = lexer.lex()
+
+        # allow for zero-length arg dicts
+        if name.kind is Kind.R_BRACE:
+            return args
+
+        if name.kind is not Kind.IDENT:
+            raise PassPipelineParseError(name, "Expected argument name here")
+
+        # handle zero-length args
+        if lexer.peek().kind in (Kind.SPACE, Kind.R_BRACE):
+            args[name.span.text] = []
+            continue
+
+        # consume the `=` in between
+        equals = lexer.lex()
+        if equals.kind is not Kind.EQUALS:
+            raise PassPipelineParseError(
+                lexer.lex(), "Expected equals as part of the pass argument here"
+            )
+
+        args[name.span.text] = _parse_arg_value(lexer)
+
+        delim = lexer.lex()
+        if delim.kind is Kind.SPACE:
+            continue
+        if delim.kind is not Kind.R_BRACE:
+            raise PassPipelineParseError(
+                delim, "Malformed pass arguments, expected either a space or `}` here"
+            )
+
+        return args
+
+
+def _parse_arg_value(lexer: PipelineLexer) -> _PassArgTypes:
+    """
+    Parse an argument value of the form: value (`,` value)*
+    """
+    elms = [_parse_arg_value_element(lexer)]
+    while lexer.peek().kind is Kind.COMMA:
+        lexer.lex()
+        elms.append(_parse_arg_value_element(lexer))
+    return elms
+
+
+def _parse_arg_value_element(lexer: PipelineLexer) -> str | int | bool | float:
```

Comment:
Maybe add an `_PassArgType` alias?

Reply:
So then if you add something to the list, it adds it to `_PassArgTypes` as well.